torch/ao/quantization/fx/_decomposed.py

```diff
@@ -94,3 +94,31 @@ def choose_qparams_tensor(input, quant_min, quant_max, dtype):
     observer(input)
     scale, zero_point = observer.calculate_qparams()
     return (scale, zero_point)
+
+# Helper function used to implement per-channel quantization against any axis
+def _permute_to_axis_zero(x, axis):
+    new_axis_list = list(range(x.dim()))
+    new_axis_list[axis] = 0
+    new_axis_list[0] = axis
+    y = x.permute(tuple(new_axis_list))
+    return y, new_axis_list
+
+quantized_decomposed_lib.define(
+    "quantize_per_channel(Tensor input, Tensor scales, Tensor zero_points, int axis, int quant_min, int quant_max, ScalarType dtype) -> Tensor")
+
+@impl(quantized_decomposed_lib, "quantize_per_channel", "CompositeExplicitAutograd")
+def quantize_per_channel(input, scales, zero_points, axis, quant_min, quant_max, dtype):
+    assert input.dtype == torch.float32, f"Expecting input to have dtype torch.float32, but got dtype: {input.dtype}"
+    _quant_min_max_bounds_check(quant_min, quant_max, dtype)
+    input, permute_axis_list = _permute_to_axis_zero(input, axis)
+    res = torch.zeros_like(input)
+
+    for i in range(input.size(0)):
+        res[i] = torch.clamp(
+            torch.round(input[i] * (1.0 / scales[i]) + zero_points[i]),
```

Comment:
I would have expected the cast to integer to happen here so the clamp is in the integer domain, is there a reason the cast is in L124?

Reply:
you mean 
```
(torch.round(input[i] * (1.0 / scales[i]) + zero_points[i]).to(dtype)
```?

I think dtype conversion need to happen after clamp is done: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtils.h#LL86, clamp here is actually happening in the floating point domain as well.

haven't compared the effect of doing this before clamping but my guess is that we want to make sure the value is within the range of dtype before we do this dtype conversion so that no out of bounds errors would happen

Reply:
makes sense, I agree with you on the cast.

It seems like the rounding should be changed to have add the zero_point after the round?  See https://github.com/pytorch/pytorch/pull/37188 .

Since it's super easy to miss these details even with test coverage (because we often aren't going to cover all of the corner cases), how about making sure this is unified in c++ land and calling the same function from all callsites?

Reply:
oh right, I just checked the code, we are using non legacy: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/AffineQuantizerBase.cpp#L51, which is doing rounding before adding zero_point: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtils.h#L78-L82
I'll change the order to align with this.

for C++, I opened this issue to track it: https://github.com/pytorch/pytorch/issues/89303, can we fix it separately? this feels like larger scope since we also have fake quant as well

Reply:
Sure, sounds reasonable to fix this later since this path is separated from the current production path.  It would be good to make sure this gets tracked and fixed before any production models are using the new path.